### ISSUE TITLE
Nice!

### ISSUE THREAD

**Issue by mzur** [jevets/kirby-form#1]
Hi Steve,

I'm the author of [Kirby Uniform](https://github.com/mzur/kirby-uniform), another form plugin. I really like your more object oriented and well structured approach. I always wanted to do a refactoring of Uniform since it "evolved" quite a bit (if you know what I mean :wink:) and, initially, I didn't really know how good PHP looks like. So I have a few questions for you:

Would you be willing to improve kirby-form to feature parity with Uniform so it can be a "better" successor? This would include CSRF and (pluggable) spam protection mechanisms as well as some analog to Uniform actions and optional multi language support. Also, it should be installable with Composer, the Kirby CLI and "standalone". If not, would you mind me borrowing some of your concepts (use Kirby\Flash, maybe extend your Form class) to develop a Uniform v3?

**Edit:** I couldn't wait and just started a little [experiment](https://github.com/mzur/kirby-uniform/tree/v3) with Uniform extending your basic Form class. I really like it. But it's nothing official yet as I'd like to hear your thoughts first.

**Comment by jevets** [comment]
Hey Martin,

Yeah, I'd be happy to combine forces on this. Thanks for reaching out!

I had checked out Uniform and really liked it, but at the time I needed to use the [POST-REDIRECT-GET paradigm](https://en.wiki2.org/wiki/Post/Redirect/Get+Milds) of dealing with forms. It helps reduce accidental (re)submits and abandonment involved when a browser asks if user would like to resubmit POST data. And it helps when dealing with validation errors and all that stuff. Then I ended up packaging it up for a few other projects.  Lots of the concepts are inspired by Laravel and its validation library, along with concepts from other frameworks.

I know exactly what you mean about how these things can evolve!

I'm totally fine with you using my classes and design on this, would be nice if we could keep `jevets\kirby-form` as its own base component, since others may want to just use it standalone. But I'm happy to work toward a v2 that provides a more stable API, along with some tests, so Uniform (and other packages?) can rely on it with more confidence.

I'll happily contribute to the development of the successor package(s), but I won't have much time until the new year. 

Sounds like fun, Martin. Thanks again for reaching out, and nice to make your acquaintance!

**Comment by mzur** [comment]
Sounds good! I've already come quite far with my little Uniform v3 experiment. It can be installed with Composer, the Kirby CLI and as traditional plugin. It validates a CSRF token, uses replacable guards for spam protection and can perform various actions. A simple example looks like this:

Controller:

```php
use Uniform\Form;
use Uniform\Actions\LogAction;
use Uniform\Actions\EmailAction;

return function ($site, $pages, $page) {
   $form = new Form([
      '_from' => [
         'rules' => ['required', 'email'],
         'message' => 'Email is required',
      ],
      'message' => [],
      'website' => [],
   ]);

   if (r::is('POST')) {
      $form->action(EmailAction::class, [
         'to' => $page->email(),
         'sender' => $site->email(),
      ])
      ->action(LogAction::class, [
         'file' => kirby()->roots()->site().DS.'form.log'
      ]);
   }

   return compact('form');
};
```

Template:

```html+php
<form method="POST">
   <input name="_from" type="email" value="<?php echo $form->old('_from'); ?>">
   <textarea name="message"><?php echo $form->old('message'); ?></textarea>
   <?php echo csrf_field(); ?>
   <?php echo honeypot_field(); ?>
   <input type="submit" value="Submit">
</form>
<?php if ($form->success()): ?>
   Success!
<?php else: ?>
   <?php snippet('uniform/errors', ['form' => $form]); ?>
<?php endif; ?>
```

When an action is executed it automatically performs the validation and runs the default "honeypot" guard (which uses a `website` field as honeypot) if no guard was called yet. If the validation, guards or actions fail, the form redirects back with error messages. What do you think of the concept?

The following issues still need to be resolved:

- You should [choose a license](http://choosealicense.com/) for kirby-flash and kirby-form. Something compatible with the GPL of Uniform would be great!

- I have to think about how multiple forms on the same page could work. If the forms don't use the same field names and the controller checks which of the forms to execute in a POST request it should work. The forms can use the same CSRF token and session information (e.g. for a captcha guard). Only the error messages might be a problem because there is no information on which form they belong to. Maybe different Flash instances with unique session keys for each form might be a solution?

- A configurable redirect URL. Currently my form redirects to `page()->url()` which might not always be desirable. Maybe set this as default and add `redirectUrl` and `withoutRedirect` methods. The redirect needs to be disabled if the form is used in an endpoint for AJAX requests. Flashing the data to the session is also not necessary in this case.

- A better `required` validator. The one of the Toolkit rejects all falsy values like `'0'` as well. This should be fixed in the Toolkit. Uniform v2 uses a pretty complex [custom version](https://github.com/mzur/kirby-uniform/blob/f139a44cb49e5ab505c3134698c7fa2dcf4cda8a/uniform/lib/UniForm.php#L156) of `a::missing` for this. ([PR](https://github.com/getkirby/toolkit/pull/205))

- A changed `Form::addErrors` method. Currently I [override](https://github.com/mzur/kirby-uniform/blob/ef69e4d1af0085669c7ed915daa94748e631fbfb/src/Form.php#L60) it so multiple error messages can be stored for the same key (similar to Laravel). This could be done in your basic Form class.

I'm quite excited to have finally gotten around to refactor Uniform :blush: I hope people will like it despite all the breaking changes. Some tests would be great, too, but my last attempt failed because of all the static methods of the Kirby Toolkit. Do you know any method how to mock these?

**Comment by mzur** [comment]
I've now released Uniform v3.0.0-beta where most of the issues above have been addressed. Only the error messages of multiple static forms still don't work but I consider this a very unusual use case.
